// graphql-schema/lib/enumeration.js
'use strict';

exports.__esModule = true;
var _bind = Function.prototype.bind;
exports['default'] = enumType;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

var _invariant = require('invariant');

var _invariant2 = _interopRequireDefault(_invariant);

var _graphql = require('graphql');

var Enumeration = (function () {
  function Enumeration(name, description) {
    _classCallCheck(this, Enumeration);

    this.name = name;
    this.description = description;
    this.values = {};
    this.lastValue = null;
  }

  Enumeration.prototype.value = function value(name, _value, description) {
    _invariant2['default'](!this.values[name], 'value(...): \'' + name + '\' is already defined');

    this.lastValue = this.values[name] = {
      value: _value,
      description: description
    };

    return this;
  };

  Enumeration.prototype.deprecated = function deprecated(deprecationReason) {
    _invariant2['default'](this.lastValue, 'deprecated(...): Deprecated must appear after a value');

    this.lastValue.deprecationReason = deprecationReason;
    return this;
  };

  Enumeration.prototype.end = function end() {
    var name = this.name;
    var description = this.description;
    var values = this.values;

    return new _graphql.GraphQLEnumType({
      name: name,
      description: description,
      values: values
    });
  };

  return Enumeration;
})();

function enumType() {
  for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
    args[_key] = arguments[_key];
  }

  return new (_bind.apply(Enumeration, [null].concat(args)))();
}

module.exports = exports['default'];
